iam_role.go

```diff
@@ -249,6 +249,117 @@ func (c *Client) CreateIamTrustRole(options *CreateIamRoleOptions) (*IamRoleResp
 	return cr, nil
 }
 
+type IamRoleInput struct {
+	RoleName                    *string            `json:"roleName"`
+	RoleType                    *string            `json:"roleType,omitempty"`
+	IncDefPols                  *int               `json:"includeDefaultPolicy,omitempty"`
+	AlksAccess                  *bool              `json:"enableAlksAccess,omitempty"`
+	TrustArn                    *string            `json:"trustArn,omitempty"`
+	TemplateFields              *map[string]string `json:"templateFields,omitempty"`
+	MaxSessionDurationInSeconds *int               `json:"maxSessionDurationInSeconds,omitempty"`
+	Tags                        *[]Tag             `json:"tags,omitempty"`
+}
+
+type IamRoleOutput struct {
+	BaseResponse
+	RoleName                    string            `json:"roleName"`
+	RoleType                    string            `json:"roleType,omitempty"`
+	RoleArn                     string            `json:"roleArn,omitempty"`
+	RoleIPArn                   string            `json:"instanceProfileArn,omitempty"`
+	RoleAddedToIP               bool              `json:"addedRoleToInstanceProfile,omitempty"`
+	Exists                      bool              `json:"roleExists,omitempty"`
+	TemplateFields              map[string]string `json:"templateFields,omitempty"`
+	MaxSessionDurationInSeconds int               `json:"maxSessionDurationInSeconds,omitempty"`
+	Tags                        *[]Tag            `json:"tags,omitempty"`
+}
+
+type requestOp struct {
+	ReqObj   []byte
+	Method   string
+	Endpoint string
+}
+
+func (input *IamRoleInput) jsonEncode(c *Client) ([]byte, error) {
+	obj, e := json.Marshal(struct {
+		IamRoleInput
+		AccountDetails
+	}{*input, c.AccountDetails})
+	if e != nil {
+		return nil, e
+	}
+	return obj, e
+}
+
+func (c *Client) newIamRoleRequest(operation *requestOp) (*IamRoleOutput, error) {
```

Comment:
The other direction you can take on this is to pass the output into this func as an interface:

```go
func (c *Client) newIamRoleRequest(operation *requestOp, respObj interface{}) error {
    if e = decodeBody(resp, &respObj); e != nil {
    }
}
```

The `decodeBody` func already takes an `interface{}`.

So, the caller now does something like this:

```
func caller() {
    respObj := &IamRoleOuptut{}
    newIamRoleRequest(op, respObj)
}
```

This would make `newIamRoleRequest` (which I still recommend renaming) completely generic and can moved into client.go (eventually, as we change all of the API calls to conform).

Reply:
I put a pointer to the response object into the request object. I figure they're dependent enough to pass together in a single object
```
req := &requestOp{
		ReqObj:   obj,
		Method:   "POST",
		Endpoint: "/updateRole/",
		Response: &IamRoleOutput{},
	}
```